lib/Doctrine/DBAL/DriverManager.php

```diff
@@ -258,53 +278,97 @@ private static function parseDatabaseUrl(array $params)
             $params['password'] = $url['pass'];
         }
 
-        if (isset($url['path'])) {
-            $params = self::parseDatabaseUrlPath($url, $params);
-        }
-
-        if (isset($url['query'])) {
-            $query = array();
-            parse_str($url['query'], $query); // simply ingest query as extra params, e.g. charset or sslmode
-            $params = array_merge($params, $query); // parse_str wipes existing array elements
-        }
+        $params = self::parseDatabaseUrlPath($url, $params);
+        $params = self::parseDatabaseUrlQuery($url, $params);
 
         return $params;
     }
 
     /**
-     * Parses the given URL and resolves the given connection parameters.
+     * Parses the given connection URL and resolves the given connection parameters.
+     *
+     * Assumes that the connection URL scheme is already parsed and resolved into the given connection parameters
+     * via {@link parseDatabaseUrlScheme}.
      *
      * @param array $url    The URL parts to evaluate.
      * @param array $params The connection parameters to resolve.
      *
      * @return array The resolved connection parameters.
+     *
+     * @see parseDatabaseUrlScheme
      */
     private static function parseDatabaseUrlPath(array $url, array $params)
     {
-        if (!isset($url['scheme'])) {
-            $params['dbname'] = $url['path'];
-
+        if (! isset($url['path'])) {
             return $params;
         }
 
-        $url['path'] = substr($url['path'], 1);
+        $url['path'] = self::normalizeDatabaseUrlPath($url['path']);
```

Comment:
True that. I don't have a strong opinion about that, my preference was to make things clear by defining each step that is part of the parsing process as individual method.